src/NetAnalyzers/Core/Microsoft.NetCore.Analyzers/Performance/PreferConvertToHexStringOverBitConverter.Fixer.cs

```diff
@@ -0,0 +1,102 @@
+            var codeAction = CodeAction.Create(
+                MicrosoftNetCoreAnalyzersResources.PreferConvertToHexStringOverBitConverterCodeFixTitle,
+                ReplaceWithConvertToHexStringCall,
+                nameof(MicrosoftNetCoreAnalyzersResources.PreferConvertToHexStringOverBitConverterCodeFixTitle));
+
+            context.RegisterCodeFix(codeAction, context.Diagnostics);
+
+            async Task<Document> ReplaceWithConvertToHexStringCall(CancellationToken cancellationToken)
+            {
+                var editor = await DocumentEditor.CreateAsync(context.Document, cancellationToken).ConfigureAwait(false);
+                var generator = editor.Generator;
+                var bitConverterArgumentsInParameterOrder = bitConverterInvocation.Arguments.GetArgumentsInParameterOrder();
+
+                var convertToHexStringSymbol = bitConverterArgumentsInParameterOrder.Length == 1
+                    ? symbols.ConvertToHexString!
+                    : symbols.ConvertToHexStringStartLength!;
+                var typeExpression = generator.TypeExpressionForStaticMemberAccess(convertToHexStringSymbol.ContainingType);
+                var methodExpression = generator.MemberAccessExpression(typeExpression, convertToHexStringSymbol.Name);
+                var methodInvocation = bitConverterArgumentsInParameterOrder.Length switch
+                {
+                    // BitConverter.ToString(data).Replace("-", "") => Convert.ToHexString(data)
+                    1 => generator.InvocationExpression(methodExpression, bitConverterArgumentsInParameterOrder[0].Value.Syntax),
+                    // BitConverter.ToString(data, start).Replace("-", "") => Convert.ToHexString(data, start, data.Length - start)
+                    2 => generator.InvocationExpression(
+                        methodExpression,
+                        bitConverterArgumentsInParameterOrder[0].Value.Syntax,
+                        bitConverterArgumentsInParameterOrder[1].Value.Syntax,
+                        generator.SubtractExpression(
+                            generator.MemberAccessExpression(
+                                bitConverterArgumentsInParameterOrder[0].Value.Syntax,
+                                WellKnownMemberNames.LengthPropertyName),
+                            bitConverterArgumentsInParameterOrder[1].Value.Syntax)),
+                    // BitConverter.ToString(data, start, length).Replace("-", "") => Convert.ToHexString(data, start, length)
+                    3 => generator.InvocationExpression(methodExpression, bitConverterArgumentsInParameterOrder.Select(a => a.Value.Syntax).ToArray()),
+                    _ => throw new NotImplementedException()
+                };
+
+                // BitConverter.ToString(data).ToLower().Replace("-", "") => Convert.ToHexString(data).ToLower()
+                if (toLowerInvocation is not null)
+                {
+                    methodInvocation = generator.InvocationExpression(
```

Comment:
At first, I tried to use the most fitting `ToLower*` method here instead of using the one in the original code, but I got many diagnostics:

- `Convert.ToHexString(data).ToLower()` triggers both CA1304 and CA1311
- `Convert.ToHexString(data).ToLowerInvariant()` triggers CA1308
- `Convert.ToHexString(data).ToLower(CultureInfo.CurrentCulture)` does not trigger a diagnostic, but seems like excessive when only converting hex characters.

So I decided to use the original `ToLower*` invocation here.

Reply:
This can be simplified when https://github.com/dotnet/runtime/issues/60393 is ready.

Reply:
> This can be simplified when https://github.com/dotnet/runtime/issues/60393 is ready.

This is now fixed, could you update?

Reply:
I've updated the fixer to use `Convert.ToHexStringLower` if it is available (>= .NET 9 Preview 1) in 0efec42ac182d00a19c54f9665d33043689a1f5d, but I had to include a temporary `ReferenceAssemblies` in the tests for now.

We should wait with merging this PR until https://github.com/dotnet/roslyn-sdk/pull/1149 is ready.

Reply:
I've removed the temporary reference assembly in f3fee4aec29fa3e688245c44f9736559fd493573.